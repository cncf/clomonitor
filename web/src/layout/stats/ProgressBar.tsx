import getCategoryColor from '../../utils/getCategoryColor';
import styles from './ProgressBar.module.css';

interface Props {
  title: string;
  icon: JSX.Element;
  value: number;
}

const ProgressBar = (props: Props) => {
  const color = getCategoryColor(props.value);

  return (
    <div className="d-flex flex-column">
      <div className={`d-flex flex-row align-items-center mb-1 mb-md-0 ${styles.progressTitle}`}>
        <div className={`me-2 position-relative ${styles.icon}`}>{props.icon}</div>
        <div>{props.title}</div>
      </div>
      <div className="d-flex flex-row mb-3 align-items-center">
        <div className={`flex-grow-1 ${styles.progressbarWrapper}`}>
          <div className={`progress rounded-0 ${styles.progress}`}>
            <div
              className={`progress-bar ${styles.progressbar}`}
              role="progressbar"
              style={{ width: `${props.value || 1}%`, backgroundColor: `var(--rm-${color})` }}
            />
          </div>
        </div>
        <div className={`ps-1 ps-md-3 lh-1 text-end ${styles.scoreWrapper}`}>
          <small className="fw-bold lightText">{props.value}%</small>
        </div>
      </div>
    </div>
  );
};

export default ProgressBar;
