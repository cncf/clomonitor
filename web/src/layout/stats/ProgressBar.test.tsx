import { render, screen } from '@testing-library/react';

import ProgressBar from './ProgressBar';

const defaultProps = {
  title: 'Security',
  icon: <>icon</>,
  value: 75,
};

describe('ProgressBar', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<ProgressBar {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders bar', () => {
      render(<ProgressBar {...defaultProps} />);

      expect(screen.getByText('Security')).toBeInTheDocument();
      expect(screen.getByText('icon')).toBeInTheDocument();
      expect(screen.getByText('75%')).toBeInTheDocument();

      const progressBar = screen.getByRole('progressbar');
      expect(progressBar).toBeInTheDocument();
      expect(progressBar).toHaveStyle({ backgroundColor: 'var(--rm-green)' });
      expect(progressBar).toHaveStyle({ width: '75%' });
    });

    it('renders bar - 20', () => {
      render(<ProgressBar {...defaultProps} value={20} />);

      expect(screen.getByText('20%')).toBeInTheDocument();

      const progressBar = screen.getByRole('progressbar');
      expect(progressBar).toBeInTheDocument();
      expect(progressBar).toHaveStyle({ backgroundColor: 'var(--rm-red)' });
      expect(progressBar).toHaveStyle({ width: '20%' });
    });

    it('renders bar - 40', () => {
      render(<ProgressBar {...defaultProps} value={40} />);

      expect(screen.getByText('40%')).toBeInTheDocument();

      const progressBar = screen.getByRole('progressbar');
      expect(progressBar).toBeInTheDocument();
      expect(progressBar).toHaveStyle({ backgroundColor: 'var(--rm-orange)' });
      expect(progressBar).toHaveStyle({ width: '40%' });
    });

    it('renders bar - 60', () => {
      render(<ProgressBar {...defaultProps} value={60} />);

      expect(screen.getByText('60%')).toBeInTheDocument();

      const progressBar = screen.getByRole('progressbar');
      expect(progressBar).toBeInTheDocument();
      expect(progressBar).toHaveStyle({ backgroundColor: 'var(--rm-yellow)' });
      expect(progressBar).toHaveStyle({ width: '60%' });
    });

    it('renders bar - 80', () => {
      render(<ProgressBar {...defaultProps} value={80} />);

      expect(screen.getByText('80%')).toBeInTheDocument();

      const progressBar = screen.getByRole('progressbar');
      expect(progressBar).toBeInTheDocument();
      expect(progressBar).toHaveStyle({ backgroundColor: 'var(--rm-green)' });
      expect(progressBar).toHaveStyle({ width: '80%' });
    });
  });
});
