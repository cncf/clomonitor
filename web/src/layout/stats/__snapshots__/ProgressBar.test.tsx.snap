// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column"
  >
    <div
      class="d-flex flex-row align-items-center mb-1 mb-md-0 progressTitle"
    >
      <div
        class="me-2 position-relative icon"
      >
        icon
      </div>
      <div>
        Security
      </div>
    </div>
    <div
      class="d-flex flex-row mb-3 align-items-center"
    >
      <div
        class="flex-grow-1 progressbarWrapper"
      >
        <div
          class="progress rounded-0 progress"
        >
          <div
            class="progress-bar"
            role="progressbar"
            style="width: 75%;"
          />
        </div>
      </div>
      <div
        class="ps-1 ps-md-3 lh-1 text-end scoreWrapper"
      >
        <small
          class="fw-bold lightText"
        >
          75%
        </small>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
